gather stats from zookeeper mntr command

jpittis opened the pull request:
Zookeeper has a tcp 4 letter command API that reports information worth logging.
Currently only the 'stat' command is being read. This adds 'mntr' command.

I'm trying to follow similar style to the parsing of the 'stat' command. Currently, all the results are reported as gauges with a tag of the given state. (one of 'standalone', 'leader' or 'follower')

I'm having trouble visualizing how this will be displayed in DataDog. I am also concerned with information conflicts between the 'stat' and 'mntr' command.

To test locally:
1. Make sure you have zookeeper running locally on the default port.
2. Setup the dd-agent repo as explained in the readme.
3. Add the following code block to the bottom of the `checks.d/zk.py` file.
```` python
if __name__ == '__main__':
    import json
    check, instances = ZookeeperCheck.from_yaml('/home/vagrant/src/dd-agent/conf.d/zk.yaml.example')
    for instance in instances:
        check.check(instance)
        if check.has_events():
            print(json.dumps(check.get_events(), indent=2, separators=(',',': ')))
        print(json.dumps(check.get_metrics(), indent=2, separators=(',',': ')))
````
4 . While inside the venv (with `. venv/bin/activate`) run the file.
````
PYTHONPATH=. python checks.d/zk.py
````

@Sirupsen @fw42 maybe have a quick glance at this?

fw42 commented:
why not PR this upstream?

Sirupsen commented:
@fw42 I suggested we review it internally and try it in production first, then PR upstream.

sirupsen commented on checks.d/zk.py:
```diff
@@ -96,6 +96,20 @@ def check(self, instance):
+        try:
+            mntr_out = self._send_command('mntr', *cx_args)
+        except ZKConnectionFailure:
+            self.increment('zookeeper.timeouts')
```
Perhaps tag these with `mntr`, `stat`, etc.?

fw42 replied:
connection failure doesn't necessarily mean timeout, does it?

sirupsen replied:
No, but it's what they do above as well.

jpittis replied:
@fw42 I'm not sure if it's trying to report zookeeper timeouts or datadog failure timeouts.
Just made me think though, we don't want to be duplicating the increment when both `stat` and mntr` fail because that will report 2 failures?

sirupsen commented on checks.d/zk.py:
```diff
@@ -96,6 +96,20 @@ def check(self, instance):
+            # Parse the response
+            metric_hash, state = self.parse_mntr(mntr_out)
+
+            # Write the data
```
Those comments aren't really necessary.

sirupsen commented on checks.d/zk.py:
```diff
@@ -211,3 +225,25 @@ def parse_stat(self, buf):
+            data = line.split()
+            if len(data) == 2 and data[0] == 'zk_server_state': # state is not an int
+                state = data[1].lower()
+            elif len(data) == 2:
```
You need an else to capture errors.

jpittis replied:
@Sirupsen Should I handle errors like the ones above? We don't want the entire thing to not report if just the `mntr` command fails?

sirupsen commented on checks.d/zk.py:
```diff
@@ -96,6 +96,20 @@ def check(self, instance):
+
+            # Write the data
+            for name in metric_hash:
+                self.gauge(name, metric_hash[name], tags=tags + [state])
```
They need the same format as the others. `zookeeper.<metric name>`, a simple concat should do it.

jpittis commented:
@Sirupsen changed tag to 'mode:<state>', removed some comments and now throw an exception on failed parse.

Sirupsen commented:
LGTM

Can you update the file in https://github.com/Shopify/cookbooks/pull/6892 and then we'll ship this tomorrow to all nodes?

Sirupsen commented:
Oh remember to update the comment at the top of the file too :)

wvanbergen commented:
Note: the `mntr` command can return "This server is currently not serving requests" if it fails to join the cluster. This is probably **the** most important thing to monitor.

Sirupsen commented:
Good find @wvanbergen. Another thing I'd like to see is what these stats look like for an observer.

wvanbergen commented:
The exact string to look for is `This ZooKeeper instance is not currently serving requests`

Sirupsen commented:
This should be tagged with something like `mode: disconnected/out-of-quorum/inactive`. We should also have a standalone metric like `zookeeper.mode.{inactive,leader,standalone,follower,observer}` where each node does an increment on every run, so we know how many are in each state and preferably which... maybe there's a better way to visualize it in DD

wvanbergen commented:
What I use elsewhere is a `set` metric, tagged with the different states. E.g. I am using the `kafkabridge.instances` metric, tagged with the git sha of the version that is running. This allows us to monitor rolling out new versions:

![screen shot 2015-06-03 at 7 00 08 am](https://cloud.githubusercontent.com/assets/15870/7958522/3e08a25e-09be-11e5-9594-2d6024a89842.png)

jpittis commented:
@wvanbergen Related to what you said: The code already checks `ruok` at the top of the program. How might that relate to an error from `mntr`? Also, does the `stat` command respond similarly? The admin docs don't seem very clear on this front.

jpittis commented:
Ah it seems that both `stat` and `mntr` can receive `This ZooKeeper instance is not currently serving requests` even after an `imok`.

jpittis commented:
Looks like the `parse_stat` already in the code does not handle this form and will just [raise an exception](https://github.com/Shopify/dd-agent/blob/zookeeper-mntr/checks.d/zk.py#L154) in this case. I have a feeling it will require re-writing maybe? @Sirupsen @wvanbergen

jpittis commented:
Looks to me like `stat` isn't worth parsing for DataDog. It's a subset of `mntr` except for a list of connected clients.

jpittis commented:
Removed `stat` command entirely. `mntr` is the only thing being parsed.

As suggested by @wvanbergen [zookeeper.instances](https://github.com/Shopify/dd-agent/blob/zookeeper-mntr/checks.d/zk.py#L98) now reports
```
mode:{inactive,leader,standalone,follower,observer}
```
I also updated the header comment which I suggest you read.
I don't know how to test the `This ZooKeeper instance is not currently serving requests` error.
@Sirupsen

wvanbergen commented on checks.d/zk.py:
```diff
@@ -71,22 +84,21 @@ def check(self, instance):
+            if state != 'inactive':
+                for name in metrics:
+                    self.gauge(name, metrics[name], tags=tags + [state_tag])
+            self.set('zookeeper.instances', state_tag, tags=[state_tag])
```
You should use the hostname (or something else that is unique for every node) as value, and the state as tag.

jpittis replied:
What should be used as a unique value? We could provide the hostname in the config file but it doesn't have it right now because it only accesses localhost... or am I miss understanding something.

sirupsen replied:
Use the hostname of the machine as @wvanbergen points out:

```
import socket
socket.gethostname()
```

Sirupsen commented:
@jpittis use something like smoketest (I don't know if it can do this sort of thing), Docker + iptables (something like `iptables -A INPUT -j REJECT`), or Toxiproxy.

jpittis commented:
@Sirupsen It's backwards compat again.
Basically, it:
1. runs the old stat command
2. uses that version from that command to check if it can run mntr
Values will be logged twice under different names in some cases:
Example:
`stat` reports: zookeeper.latency.avg
`mntr` reports: zookeeper.avg.latency

Why both?
`stat` naming convention needs to be kept for backwards compat
`mntr` naming convention is more standardized and should be used over the other one.

If upstream doesn't like that, easy fix to remove the duplicates.

Gonna look into testing for failure with toxi proxy or something.

Sirupsen commented:
K sounds fine to me, make sure to write that in the upstream PR.

Does this library have any tests? You don't have to write automated tests for the failure mode IMO., just prove to yourself hardcoding against Willem's string works/.*

jpittis commented:
@Sirupsen It does have tests. I'll look into making them pass once I confirm Willem's string.

jpittis commented:
@Sirupsen: I tested willem's suggestion manually and the last commit handles it.
Confident that the code is good to go and would only need possible cleaning before a upstream pull.
Don't have time to fix the tests tonight. I'll get them working tomorrow.
(don't even know if they we're passing to begin with. lots of depreciated code in this repo)

Sirupsen commented:
Yeah, I remember getting them to run at some point and it was unpleasant. Can you update the code in the cookbook PR so we can ship that tomorrow morning?

jpittis commented:
When zookeeper is down, this check would raise an exception, stopping any data being sent to DataDog.
The script will now report a `zookeeper.instances` `mode:down` when it cannot reach zookeeper.

Before, the script would not report anything to DataDog if an exception occurred. Instead, now when the script encounters an exception, `zookeeper.check.d.exception` is incremented and `zookeeper.instances` is set to `mode:unknown`.

@Sirupsen I want to pop this into the cookbook asap so we can test it before the end of the day.

Sirupsen commented:
@jpittis just make another PR to the cookbooks to update the script

jpittis commented:
@Sirupsen: This logs the exception and renames the the increment to `datadog_client_exception`.
Feels ugly. Thinking I'll refactor it tomorrow. Might be worth testing the cookbook before I head out?

Sirupsen commented:
@jpittis if you merge your PR and message me on slack I'll kill the node for you.

jpittis commented:
@camilo, @fw42 @fbogsany: I'm looking to PR this upstream. Don't know Python too well so I'm feeling a little uncomfortable. Would you guys have any last minute reviews, or suggestions regarding the process of sending this upstream.

Keep in mind that 7e632de has been running successfully in production for a good few weeks.

Thanks.

cc @Sirupsen

fw42 commented:
ping @angelini and @orenmazor for python review please

angelini commented on checks.d/zk.py:
```diff
@@ -96,6 +156,50 @@ def check(self, instance):
+            'unknown': 0
+        }
+
+        if mode in gauges.keys():
```
You don't need `keys` here

angelini commented on checks.d/zk.py:
```diff
@@ -210,4 +317,34 @@ def parse_stat(self, buf):
+        ''' `buf` is a readable file-like object
+            returns a tuple: (metrics, mode)
+            if mode == 'inactive', metrics will be None
+        '''
```
If your docstring is going to follow the Python spec, `return` should be wrapped in colons. `:return:`

https://pythonhosted.org/an_example_pypi_project/sphinx.html#function-definitions

angelini commented on checks.d/zk.py:
```diff
@@ -210,4 +317,34 @@ def parse_stat(self, buf):
+                name = data[0].replace('zk', 'zookeeper').replace('_', '.')
+                metrics[name] = data[1]
+            else:
+                raise Exception("Data not in 'key value' format, could not parse '%s'" % data.join(' '))
```
This should be `' '.join(data)`, `join` is a method on `str` and not on `list`

angelini commented on checks.d/zk.py:
```diff
@@ -96,6 +156,50 @@ def check(self, instance):
+                mode_tag = "mode:%s" % mode
+                if mode != 'inactive':
+                    for name in metrics:
+                        self.gauge(name, metrics[name], tags=tags + [mode_tag])
```
This is more of a style nitpick then a Python specific problem, but this code is adding 20 lines to a method that's already 80 lines long. It looks like this could be pulled out into it's own method

orenmazor commented:
> why not PR this upstream?

I'm with @fw42. I'd love to see this upstream asap.

jpittis commented:
This was merged upstream from d0478d6dfe62d75b4e5d2c1718b813693668d2a7 through to 68c6debd536c65b1eed690c1d9b414f7d5d7288c.

There's no priority to updating to this code. I can do the update in the beginning of May.
